lib/test_state.go

```diff
@@ -31,6 +33,118 @@ type TestRunState struct {
 	Runner  Runner // TODO: rename to something better, see type comment
 	RunTags *metrics.TagSet
 
+	GroupSummary *GroupSummary // TODO(@mstoykov): move and rename
+
 	// TODO: add other properties that are computed or derived after init, e.g.
 	// thresholds?
 }
+
+// GroupSummaryDescription is the description of the GroupSummary used to identify and ignore it
+// for the purposes of the cli descriptions.
+const GroupSummaryDescription = "Internal Group Summary output"
+
+// GroupSummary is an internal output implementation that facilitates the aggregation of
+// group and check metrics for the purposes of the end of test summary and the REST API
+type GroupSummary struct {
+	group   *Group // TODO(@mstoykov): move the whole type outside of lib later
+	ch      chan []metrics.SampleContainer
+	stopped chan struct{}
+	logger  logrus.FieldLogger
+}
+
+// NewGroupSummary returns new GroupSummary ready to be started.
+func NewGroupSummary(logger logrus.FieldLogger) *GroupSummary {
+	group, _ := NewGroup(RootGroupPath, nil)
+	return &GroupSummary{
+		group:   group,
+		ch:      make(chan []metrics.SampleContainer, 1000),
+		stopped: make(chan struct{}),
+		logger:  logger,
+	}
+}
+
+// Group returns the underlying group that has been aggregated
+func (gs *GroupSummary) Group() *Group {
+	return gs.group
+}
+
+// Description is part of the output.Output interface
+func (gs *GroupSummary) Description() string {
+	return GroupSummaryDescription
+}
+
+func buildGroup(rootGroup *Group, groupName string) (*Group, error) {
+	group := rootGroup
+	groups := strings.Split(groupName, "::")
+	var err error
+	for _, groupName := range groups[1:] {
+		group, err = group.Group(groupName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return group, nil
+}
+
+func (gs *GroupSummary) handleSample(sample metrics.Sample) error {
+	switch sample.Metric.Name {
+	case "group_duration", "checks":
+		groupName, ok := sample.Tags.Get(metrics.TagGroup.String())
+		if !ok {
+			return nil
+		}
+		group, err := buildGroup(gs.group, groupName)
+		if err != nil {
+			return err
+		}
+
+		if sample.Metric.Name != "checks" {
```

Comment:
```suggestion
		// In case of checks we need to also generate the check and add it values. Otherwise we are done.
		if sample.Metric.Name != "checks" {
```